src/arraytraits.rs

```diff
@@ -94,6 +94,46 @@ impl<S, D, I> IndexMut<I> for ArrayBase<S, D>
     }
 }
 
+impl<'a, A, D> ArrayView<'a, A, D>
+where
+    D: Dimension,
+{
+    /// Get a reference of a element through the view.
+    ///
+    /// This is a replacement of `Index::index` since it forces us a wrong lifetime
+    /// https://github.com/bluss/rust-ndarray/issues/371
+    pub fn elem_ref<I: NdIndex<D>>(&self, index: I) -> &'a A {
+        debug_bounds_check!(self, index);
+        unsafe {
+            &*self.as_ptr().offset(
+                index
+                    .index_checked(&self.dim, &self.strides)
+                    .unwrap_or_else(|| array_out_of_bounds()),
+            )
+        }
+    }
+}
+
+impl<'a, A, D> ArrayViewMut<'a, A, D>
+where
+    D: Dimension,
+{
+    /// Get a mutable reference of a element through the view.
+    ///
+    /// This is a replacement of `IndexMut::index_mut` since it forces us a wrong lifetime
+    /// https://github.com/bluss/rust-ndarray/issues/371
+    pub fn elem_ref_mut<I: NdIndex<D>>(&mut self, index: I) -> &'a mut A {
```

Comment:
See the comment in the PR thread first. We'll have to be creative here, I don't know what we can do that is also correct with Rust's borrow checker rules.

(As a digression, it's a shame we have this problem! Slices are special in Rust, they are built in dynamically sized types. We'd like to have a custom version of that instead of what we have with the ArrayView and ArrayViewMut. We wouldn't have this problem.)